sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1003 @@
+                # fitting in memory
+                ii, jj = self._find_impostors_batch(Lx[ind_out], Lx[ind_in],
+                                                    margin_radii[ind_out],
+                                                    margin_radii[ind_in])
+
+                if len(ii):
+                    # sample constraints if they are too many
+                    if len(ii) > self.max_constraints:
+                        dims = (len(ind_out), len(ind_in))
+                        ind = np.ravel_multi_index((ii, jj), dims=dims)
+                        ind_sampled = choice(ind, self.max_constraints,
+                                             replace=False,
+                                             random_state=self.random_state_)
+                        ii, jj = np.unravel_index(ind_sampled, dims=dims)
+
+                    imp_row = ind_out[ii]
+                    imp_col = ind_in[jj]
+                    new_imp = csr_matrix((np.ones(len(imp_row), dtype=np.int8),
+                                          (imp_row, imp_col)), dtype=np.int8,
+                                         shape=(n_samples, n_samples))
+                    impostors_sp = impostors_sp + new_imp
+
+            impostors_sp = impostors_sp.tocoo(copy=False)
+            imp_row = impostors_sp.row
+            imp_col = impostors_sp.col
+            dist = pairs_distances_batch(Lx, imp_row, imp_col)
+        else:
+            # Initialize impostors vectors
+            imp_row, imp_col, dist = [], [], []
+            for class_id in self.classes_non_singleton_[:-1]:
+                ind_in, = np.where(np.equal(y, class_id))
+                ind_out, = np.where(np.greater(y, class_id))
+
+                # Subdivide ind_out x ind_in to chunks of a size that is
+                # fitting in memory
+                ii, jj, dd = self._find_impostors_batch(
+                    Lx[ind_out], Lx[ind_in], margin_radii[ind_out],
+                    margin_radii[ind_in], True)
+
+                if len(ii):
+                    # sample constraints if they are too many
+                    if len(ii) > self.max_constraints:
+                        dims = (len(ind_out), len(ind_in))
+                        ind = np.ravel_multi_index((ii, jj), dims=dims)
+                        ind_sampled = choice(len(ind), self.max_constraints,
+                                             replace=False,
+                                             random_state=self.random_state_)
+                        dd = np.asarray(dd)[ind_sampled]
+                        ind_sampled = ind[ind_sampled]
+                        ii, jj = np.unravel_index(ind_sampled, dims=dims)
+
+                    imp_row.extend(ind_out[ii])
+                    imp_col.extend(ind_in[jj])
+                    dist.extend(dd)
+
+            imp_row, imp_col = np.asarray(imp_row), np.asarray(imp_col)
+            dist = np.asarray(dist)
+
+        return imp_row, imp_col, dist
+
+    @staticmethod
```

Comment:
I there a reason that this is a method? If it doesn't access self, I would prefer it to be a function (it's more explicit that it doesn't need self), unless it is for a subclassing mechanism.